core-bundle/src/Migration/AbstractRecordedMigration.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Migration;
+
+use Contao\CoreBundle\Entity\Migration as MigrationEntity;
+use Doctrine\DBAL\Connection;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Contracts\Service\ServiceSubscriberInterface;
+use Symfony\Contracts\Service\ServiceSubscriberTrait;
+
+abstract class AbstractRecordedMigration extends AbstractMigration implements ServiceSubscriberInterface
+{
+    use ServiceSubscriberTrait;
+
+    public function shouldRun(): bool
+    {
+        return false === $this->hasRun();
+    }
+
+    protected function hasRun(): ?bool
+    {
+        if (!$this->connection()->getSchemaManager()->tablesExist(['tl_migration'])) {
+            return null;
+        }
+
+        return null !== $this->entityManager()->getRepository(MigrationEntity::class)->findOneBy(['name' => $this->getName()]);
+    }
+
+    protected function createResult(bool $successful, string $message = null): MigrationResult
+    {
+        if (false === $this->hasRun()) {
```

Comment:
Initially I also checked for `$successful` here - but that might lead to infinite executions of a migration.